### ISSUE TITLE
duplicate key name while adding indexes and constraints in db_schema.xml

### ISSUE THREAD

**Issue by ioweb-gr** [magento/magento2#27899]
Expanding on #27829 I bumped into a bug here that I can't find a way to work around. 

### Preconditions (*)
1. Magento 2.3.4

### Steps to reproduce (*)
1. Create a new module 
2. In db_schema.xml add the following declarations in a new table

 ```xml
   <table name="mytable" resource="default" engine="innodb">
        <column xsi:type="int" name="entity_id" identity="true" unsigned="true" nullable="false" comment="Entity Id" />
        <column xsi:type="varchar" length="100" name="identifier" nullable="false" comment="Identifier" />
        <column xsi:type="timestamp" name="created_at" nullable="false" comment="Creation Time" default="CURRENT_TIMESTAMP" on_update="false" />
        <column xsi:type="timestamp" name="updated_at" nullable="false" comment="Update Time" default="CURRENT_TIMESTAMP" on_update="true" />
        <!-- // Add more static attributes here...-->
        <constraint xsi:type="primary" referenceId="PRIMARY">
            <column name="entity_id" />
        </constraint>
        <index referenceId="IDENTIFIER" indexType="btree">
            <column name="identifier" />
        </index>
    </table>
```

3. Execute bin/magento setup:db-schema:upgrade
4. Modify XML like this

```xml
    <table name="mytable" resource="default" engine="innodb">
        <column xsi:type="int" name="entity_id" identity="true" unsigned="true" nullable="false" comment="Entity Id" />
        <column xsi:type="varchar" length="100" name="identifier" nullable="false" comment="Identifier" />
        <column xsi:type="timestamp" name="created_at" nullable="false" comment="Creation Time" default="CURRENT_TIMESTAMP" on_update="false" />
        <column xsi:type="timestamp" name="updated_at" nullable="false" comment="Update Time" default="CURRENT_TIMESTAMP" on_update="true" />
        <!-- // Add more static attributes here...-->
        <constraint xsi:type="primary" referenceId="PRIMARY">
            <column name="entity_id" />
        </constraint>
        <index referenceId="IDENTIFIER" indexType="btree">
            <column name="identifier" />
        </index>
        <constraint xsi:type="unique" referenceId="UNIQUE">
           <column name="identifier" />
        </constraint>
    </table>
```
5. Execute bin/magento setup:db-schema:upgrade

### Expected result (*)
1. The indexes are created 

### Actual result (*)
1. 
`  SQLSTATE[42000]: Syntax error or access violation: 1061 Duplicate key name 'MYTABLE_IDENTIFIER'`

2.
` SQLSTATE[42000]: Syntax error or access violation: 1061 Duplicate key name 'MYTABLE_IDENTIFIER', query w
  as: ALTER TABLE `mytable` ADD INDEX `MYTABLE_IDENTIFIER` (`identifier`)
`

Basically Magento is unable to generate a unique name for the index and the constraint.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @shikhamis11 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-33558` were created

**Issue Available**: @shikhamis11, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by madhurajawat** [comment]
@magento-engcom-team I am unable to assign this task. Can you please give access to assign the task

**Comment by shikhamis11** [comment]
@madhurajawat once you will create Pull Request for this issue you will be automatically assign to this issue

**Comment by madhurajawat** [comment]
> @madhurajawat once you will create Pull Request for this issue you will be automatically assigned to this issue

@shikhamis11 Ok, Thanks

**Comment by madhurajawat** [comment]
@ioweb-gr  You can add one index for a column at a time. You can either add INDEX or UNIQUE for a column. 

For adding UNIQUE constraint You have to first drop index then add a UNIQUE constraint fo that column
The query will be like below:

ALTER TABLE `mytable` DROP INDEX `MYTABLE_IDENTIFIER`, ADD UNIQUE `MYTABLE_IDENTIFIER` (`identifier`) USING BTREE; 
![image](https://user-images.githubusercontent.com/59955325/80364284-45c7f300-88a3-11ea-9451-e3a611d44cda.png)


**With the help of disabled=" true" You can add UNIQUE constraint.** 

 ```
  <table name="mytable" resource="default" engine="innodb">
       <column xsi:type="int" name="entity_id" identity="true" unsigned="true" nullable="false" comment="Entity Id" />
       <column xsi:type="varchar" length="100" name="identifier" nullable="false" comment="Identifier" />
       <column xsi:type="timestamp" name="created_at" nullable="false" comment="Creation Time" default="CURRENT_TIMESTAMP" on_update="false" />
       <column xsi:type="timestamp" name="updated_at" nullable="false" comment="Update Time" default="CURRENT_TIMESTAMP" on_update="true" />
       <!-- // Add more static attributes here...-->
       <constraint xsi:type="primary" referenceId="PRIMARY">
           <column name="entity_id" />
       </constraint>
       <index referenceId="IDENTIFIER" indexType="btree" disabled="true">
           <column name="identifier" />
       </index>
        <constraint xsi:type="unique" referenceId="UNIQUE">
           <column name="identifier" />
        </constraint>
   </table>
```

Or you can remove index row and add UNIQUE constraint row So index will be removed and UNIQUE constraint will be added.

Remove below code: 
<index referenceId="IDENTIFIER" indexType="btree">
            <column name="identifier" />
</index>

Add below code:

<constraint xsi:type="unique" referenceId="UNIQUE">
           <column name="identifier" />
 </constraint>

**Comment by ioweb-gr** [comment]
Yes exactly that's the point of the bug. While in mariadb you can define both indexes at the same time with a different name, with db_schema.xml you can't because it's reusing the same referenceId, it can't create a unique one for both at the same time because the algorithm that creates the names cannot create a unique name for both.

**Comment by engcom-Echo** [comment]
> @magento-engcom-team I am unable to assign this task. Can you please give access to assign the task

Hi @madhurajawat.
https://github.com/magento/magento2/wiki/Working-Issues-and-PRs

**Comment by wutever0** [comment]
facing the same issue. Would make sense if the SQL generated would use the referenceId value as the name of the key.

**Comment by ilnytskyi** [comment]
Faced this issue when creating constraint and index for two columns
Workaround: Change order of field in `constraint` or `index`
```
        <constraint xsi:type="unique" referenceId="ID_TYPE">
            <column name="identifier"/>
            <column name="type"/>
        </constraint>
        <index referenceId="TYPE_ID" indexType="btree">
            <column name="type"/>
            <column name="identifier"/>
        </index>
```
The same order will generate an error

**Comment by gdvisser** [comment]
If you are facing this issue when creating both a unique constraint and index for specific column(s) you can remove the index from the schema. The unique constraint will already create an (unique) index on the column(s).

**Comment by dunagan5887** [comment]
@gdvisser  Do you have documentation from MySQL to support that?

**Comment by dverkade** [comment]
Confirmed issues can't be closed automatically.

**Comment by aneneve-corra** [comment]
I know that this was so long ago, but if someone comes into this same issue, here's a patch file: 
[27899_magento-framework.patch](https://github.com/magento/magento2/files/11105124/27899_magento-framework.patch)
